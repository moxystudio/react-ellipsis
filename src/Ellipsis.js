import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { throttle } from 'lodash';

class Ellipsis extends Component {
    containerRef = createRef();

    state = {
        isPrepared: false,
    };

    constructor(props) {
        super(props);

        this.throttleHandleResize = throttle(this.handleResize, 250);
    }

    componentDidMount() {
        window.addEventListener('resize', this.throttleHandleResize, { passive: true });

        this.ellipsize();

        this.setState({ isPrepared: true });
    }

    componentDidUpdate() {
        this.ellipsize();
    }

    componentWillUnmount() {
        window.removeEventListener('resize', this.throttleHandleResize, { passive: true });
    }

    render() {
        const { children, tag: Tag, className } = this.props;
        const { isPrepared } = this.state;

        const containerClassName = classNames('container', {
            reveal: isPrepared,
        }, className);

        return (
            <Tag
                ref={ this.containerRef }
                className={ containerClassName }
                { ...this.props }>
                { children }
            </Tag>
        );
    }

    ellipsize = () => {
        const element = this.containerRef.current;
        const words = this.props.children;
        let wordsArray = [];

        element.innerHTML = words;

        wordsArray = typeof words === 'string' && words.split(' ');

        /* istanbul ignore next */
        while (element.scrollHeight > element.offsetHeight && wordsArray.length > 0) {
            wordsArray.pop();
            element.innerHTML = `${wordsArray.join(' ')}...`;
        }
    };

    /* istanbul ignore next */
    handleResize = () => {
        this.ellipsize();
    };
}

Ellipsis.propTypes = {
    children: PropTypes.string.isRequired,
    tag: PropTypes.string,
    className: PropTypes.string,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,
};

Ellipsis.defaultProps = {
    tag: 'p',
};

export default Ellipsis;
