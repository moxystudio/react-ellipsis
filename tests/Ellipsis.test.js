import React from 'react';
import { render } from '@testing-library/react';

import Ellipsis from '../src/Ellipsis';

const defaultProps = {
    children: 'Testing text, testing text.',
    className: 'className',
};

const renderWithProps = (props = {}) => render(<Ellipsis { ...defaultProps } { ...props } />);

describe('Ellipsis Component', () => {
    it('should render correctly', () => {
        const { asFragment } = renderWithProps();

        expect(asFragment()).toMatchSnapshot();
    });
});
