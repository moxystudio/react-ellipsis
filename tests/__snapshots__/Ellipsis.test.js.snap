// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ellipsis Component should render correctly 1`] = `
<DocumentFragment>
  <p
    class="react-ellipsis_container reveal className"
  >
    Testing text, testing text.
  </p>
</DocumentFragment>
`;

exports[`Ellipsis Component should render correctly when there is a resize 1`] = `
<DocumentFragment>
  <p
    class="react-ellipsis_container reveal className"
  >
    Testing text, testing text.
  </p>
</DocumentFragment>
`;
